code/modules/mob/living/basic/space_fauna/revenant/_revenant.dm

```diff
@@ -0,0 +1,513 @@
+	var/max_essence = 75
+	/// If the revenant regenerates essence or not
+	var/essence_regenerating = TRUE
+	/// How much essence regenerates per second
+	var/essence_regen_amount = 2.5
+	/// How much essence the revenant has stolen
+	var/essence_accumulated = 0
+	/// How much stolen essence is available for unlocks
+	var/essence_excess = 0
+	/// How long the revenant is revealed for, is about 2 seconds times this var.
+	var/unreveal_time = 0
+	/// How long the revenant is stunned for, is about 2 seconds times this var.
+	var/unstun_time = 0
+	/// How many perfect, regen-cap increasing souls the revenant has. //TODO, add objective for getting a perfect soul(s?)
+	var/perfectsouls = 0
+
+/mob/living/basic/revenant/Initialize(mapload)
+	. = ..()
+	AddElement(/datum/element/simple_flying)
+	add_traits(list(TRAIT_SPACEWALK, TRAIT_SIXTHSENSE, TRAIT_FREE_HYPERSPACE_MOVEMENT), INNATE_TRAIT)
+
+	// Starting spells
+
+	var/datum/action/cooldown/spell/list_target/telepathy/revenant/telepathy = new(src)
+	telepathy.Grant(src)
+
+	// Starting spells that start locked
+	var/datum/action/cooldown/spell/aoe/revenant/overload/lights_go_zap = new(src)
+	lights_go_zap.Grant(src)
+
+	var/datum/action/cooldown/spell/aoe/revenant/defile/windows_go_smash = new(src)
+	windows_go_smash.Grant(src)
+
+	var/datum/action/cooldown/spell/aoe/revenant/blight/botany_go_mad = new(src)
+	botany_go_mad.Grant(src)
+
+	var/datum/action/cooldown/spell/aoe/revenant/malfunction/shuttle_go_emag = new(src)
+	shuttle_go_emag.Grant(src)
+
+	var/datum/action/cooldown/spell/aoe/revenant/haunt_object/toolbox_go_bonk = new(src)
+	toolbox_go_bonk.Grant(src)
+
+	RegisterSignal(src, COMSIG_LIVING_BANED, PROC_REF(on_baned))
+	RegisterSignal(src, COMSIG_MOVABLE_PRE_MOVE, PROC_REF(on_move))
+	RegisterSignal(src, COMSIG_LIVING_LIFE, PROC_REF(on_life))
+	set_random_revenant_name()
+
+	GLOB.revenant_relay_mobs |= src
+
+/mob/living/basic/revenant/Destroy()
+	GLOB.revenant_relay_mobs -= src
+	return ..()
+
+/mob/living/basic/revenant/Login()
+	. = ..()
+	if(!. || isnull(client))
+		return FALSE
+
+	var/static/cached_string = null
+	if(isnull(cached_string))
+		cached_string = examine_block(jointext(create_login_string(), "\n"))
+
+	to_chat(src, cached_string, type = MESSAGE_TYPE_INFO)
+
+	if(generated_objectives_and_spells)
+		return TRUE
+
+	generated_objectives_and_spells = TRUE
+	mind.set_assigned_role(SSjob.GetJobType(/datum/job/revenant))
+	mind.special_role = ROLE_REVENANT
+	SEND_SOUND(src, sound('sound/effects/ghost.ogg'))
+	mind.add_antag_datum(/datum/antagonist/revenant)
+	return TRUE
+
+/// Signal Handler Injection to handle Life() stuff for revenants
+/mob/living/basic/revenant/proc/on_life(seconds_per_tick = SSMOBS_DT, times_fired)
+	SIGNAL_HANDLER
+
+	if(dormant)
+		return COMPONENT_LIVING_CANCEL_LIFE_PROCESSING
+
+	if(revealed && essence <= 0)
+		INVOKE_ASYNC(src, PROC_REF(death))
+		return COMPONENT_LIVING_CANCEL_LIFE_PROCESSING
+
+	if(unreveal_time && world.time >= unreveal_time)
```

Comment:
What do you think about just making "Revenant Revealed" and "Revenant Stunned" into status effects? They can handle their own duration and application/removal effects.

Reply:
i think that's more bulk to an already bulky refactor. would be nice but sheesh it's a huge PR already

Reply:
Agree / I made them status effects when I last attempted to work on this code

Reply:
i think it's probably more worthwhile to do status effects so we aren't so reliant on Life() here so i'll look at it. i hope you like coder sprites